_deploy/r/gnoswap/gnft/gnft_test.gno

```diff
@@ -0,0 +1,450 @@
+		shouldPanic bool
+		panicMsg    string
+		expected    std.Address
+	}{
+		{"OwnerOf(1)", 1, false, "", addr01},
+		{"OwnerOf(2)", 2, true, errInvalidTokenId},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.shouldPanic {
+				uassert.PanicsWithMessage(t, tt.panicMsg, func() {
+					OwnerOf(tid(tt.tokenId))
+				})
+			} else {
+				uassert.Equal(t, tt.expected, OwnerOf(tid(tt.tokenId)))
+			}
+		})
+	}
+}
+
+func TestIsApprovedForAll(t *testing.T) {
+	tests := []struct {
+		name     string
+		setup    func()
+		expected bool
+	}{
+		{
+			name:     "IsApprovedForAll(addr01, addr02)",
+			expected: false,
+		},
+		{
+			name: "IsApprovedForAll(addr01, addr02) after setting approval",
+			setup: func() {
+				std.TestSetRealm(addr01Realm)
+				SetApprovalForAll(a2u(addr02), true)
+			},
+			expected: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.setup != nil {
+				tt.setup()
+			}
+			uassert.Equal(t, tt.expected, IsApprovedForAll(a2u(addr01), a2u(addr02)))
+		})
+	}
+}
+
+func TestGetApproved(t *testing.T) {
+	tests := []struct {
+		name             string
+		setup            func()
+		expectedAddr     std.Address
+		expectedApproved bool
+	}{
+		{
+			name:             "GetApproved(1)",
+			expectedAddr:     std.Address(""),
+			expectedApproved: false,
+		},
+		{
+			name: "GetApproved(1) after approving",
+			setup: func() {
+				std.TestSetRealm(addr01Realm)
+				Approve(a2u(addr02), tid(1))
+			},
+			expectedAddr:     addr02,
+			expectedApproved: true,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.setup != nil {
+				tt.setup()
+			}
+
+			addr, approved := GetApproved(tid(1))
+			uassert.Equal(t, tt.expectedAddr, addr)
+			uassert.Equal(t, tt.expectedApproved, approved)
+		})
+	}
+}
+
+func TestTransferFrom(t *testing.T) {
+	resetObject(t)
+
+	tests := []struct {
+		name        string
+		setup       func()
+		callerRealm std.Realm
+		shouldPanic bool
+		panicMsg    string
+		expected    std.Address
+	}{
+		{
+			name:        "transfer non-existent token id",
+			callerRealm: std.NewUserRealm(addr01),
+			shouldPanic: true,
+			panicMsg:    "caller is not token owner or approved",
+			expected:    std.Address(""),
+		},
+		{
+			name: "transfer token owned by other user",
+			setup: func() {
+				std.TestSetRealm(positionRealm)
+				Mint(a2u(addr01), tid(1))
+			},
+			callerRealm: std.NewUserRealm(addr02),
+			shouldPanic: true,
+			panicMsg:    "caller is not token owner or approved",
+			expected:    std.Address(""),
+		},
+		{
+			name:        "transfer token owned by caller",
+			callerRealm: std.NewUserRealm(addr01),
+			shouldPanic: false,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.setup != nil {
+				tt.setup()
+			}
+
+			if tt.shouldPanic {
+				uassert.PanicsWithMessage(t, tt.panicMsg, func() {
+					TransferFrom(a2u(addr01), a2u(addr02), tid(1))
```

Comment:
After the TransferFrom, you should also see additional assets deleted from the TokenList.

Reply:
2e25c41db0c6c37365eed4918f2ad05521b8c61f